components/templates/src/global_fns/load_data.rs

```diff
@@ -0,0 +1,403 @@
+extern crate toml;
+extern crate serde_json;
+
+use utils::fs::{read_file, is_path_in_directory, get_file_time};
+
+use std::hash::{Hasher, Hash};
+use std::fmt;
+use std::collections::hash_map::DefaultHasher;
+use reqwest::{Client, header};
+use url::Url;
+
+use std::path::PathBuf;
+use std::sync::{Arc, Mutex};
+
+
+use csv::Reader;
+use std::collections::HashMap;
+use tera::{GlobalFn, Value, from_value, to_value, Result, Map};
+
+static GET_DATA_ARGUMENT_ERROR_MESSAGE: &str = "`load_data`: requires EITHER a `path` or `url` argument";
+
+enum DataSource {
+    Url(Url),
+    Path(PathBuf)
+}
+
+#[derive(Debug)]
+enum OutputFormat {
+    TOML,
+    JSON,
+    CSV,
+    Plain
+}
+
+impl fmt::Display for OutputFormat {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::Debug::fmt(self, f)
+    }
+}
+
+impl Hash for OutputFormat {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.to_string().hash(state);
+    }
+}
+
+impl OutputFormat {
+    fn from(output_format: String) -> Result<Self> {
```

Comment:
you could impl [FromStr](https://doc.rust-lang.org/std/str/trait.FromStr.html) instead

Reply:
I did think of that, but because it's possible for it to fail (ie the string doesnt match one of the enum values), it's unsafe, and `TryFrom` is only in nightly right now.

Reply:
`FromStr` returns a `Result` though? `FromStr` and `TryFrom` look to be identical for this usecase

Reply:
Aha, I found my confusion. I was looking at `From` rather than `FromStr`. Yes you're right, `FromStr` would be ideal to implement here!